### PR TITLE
feat(ton): add single nominator withdraw txn

### PR DESCRIPTION
**Pull request by quarterdill**
Ticket: SC-370

Add TON single nominator withdraw transaction

### CODE REVIEW

File: modules/sdk-coin-ton/src/lib/transactionBuilderFactory.ts [hunk 1]
```diff
@@ -19,6 +20,13 @@ export class TransactionBuilderFactory extends BaseTransactionBuilderFactory {
     return new TransferBuilder(this._coinConfig);
   }
 
+  /**
+   * Returns a specific builder to create a TON withdraw transaction
+   */
+  getWithdrawBuilder(): WithdrawBuilder {
```
Comment (ewangbitgo) [inline modules/sdk-coin-ton/src/lib/transactionBuilderFactory.ts:26]: can you change the name to SingleNominatorWithdrawBuilder. Withdraw is too broad

File: modules/sdk-coin-ton/src/lib/withdrawBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  send(recipient: Recipient): WithdrawBuilder {
+    this.transaction.recipient = recipient;
+    return this;
+  }
+
+  private createMessage(): string {
+    const opcode = '00001000';
+    const query_id = '0000000000000000';
+    const amount = TonWeb.utils.toNano(this.transaction.recipient.amount);
+    return opcode.concat(query_id).concat(amount.toString('hex'));
```
Comment (ewangbitgo) [inline modules/sdk-coin-ton/src/lib/withdrawBuilder.ts:24]: this won't work. Ton is using bag of cell(boc) packing, I think you need to put it in boc structure.

File: modules/sdk-coin-ton/src/lib/withdrawBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  }
+
+  setMessage(): WithdrawBuilder {
+    this.transaction.message = this.createMessage();
```
Comment (ewangbitgo) [inline modules/sdk-coin-ton/src/lib/withdrawBuilder.ts:28]: are you sure it is just message?

File: modules/sdk-coin-ton/test/unit/withdrawBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+  it('should build a send from rawTx', async function () {
+    const txBuilder = factory.from(testData.signedTransaction.tx);
+    const builtTx = await txBuilder.build();
+    const jsonTx = builtTx.toJson();
+    should.equal(builtTx.type, TransactionType.Send);
+    should.equal(builtTx.signablePayload.toString('base64'), testData.signedTransaction.signable);
+    should.equal(builtTx.id, testData.signedTransaction.txId);
+    const builder2 = factory.from(builtTx.toBroadcastFormat());
+    const builtTx2 = await builder2.build();
+    should.equal(builtTx2.type, TransactionType.Send);
```
Comment (ewangbitgo) [inline modules/sdk-coin-ton/test/unit/withdrawBuilder.ts:86]: the type should be different from send

Comment (quarterdill) [reply modules/sdk-coin-ton/test/unit/withdrawBuilder.ts]: made it `StakingWithdraw`

File: modules/sdk-coin-ton/src/lib/withdrawBuilder.ts (hunk 3 above)
Comment (quarterdill) [reply modules/sdk-coin-ton/src/lib/withdrawBuilder.ts]: yup seems like we just need to send a message with the right body, which has the transaction hex

File: modules/sdk-coin-ton/src/lib/transactionBuilderFactory.ts [hunk 5]
```diff
@@ -1,24 +1,46 @@
...
 export class TransactionBuilderFactory extends BaseTransactionBuilderFactory {
   constructor(_coinConfig: Readonly<CoinConfig>) {
     super(_coinConfig);
   }
   /** @inheritdoc */
-  from(raw: string): TransactionBuilder {
-    const builder = new TransferBuilder(this._coinConfig);
-    builder.from(raw);
-    return builder;
+  from(raw: string, type: TransactionType = TransactionType.Send): TransactionBuilder {
```
Comment (ewangbitgo) [inline modules/sdk-coin-ton/src/lib/transactionBuilderFactory.ts:12]: you can't change the parameter list from inherited method

File: modules/sdk-coin-ton/src/lib/transaction.ts [hunk 6]
```diff
@@ -67,8 +67,8 @@ export class Transaction extends BaseTransaction {
     return Buffer.from(this.unsignedMessage, 'hex');
   }
 
-  async build(): Promise<void> {
-    this._type = TransactionType.Send;
+  async build(txType: TransactionType): Promise<void> {
```
Comment (ewangbitgo) [inline modules/sdk-coin-ton/src/lib/transaction.ts:70]: you shouldn't change the build method signature here, the type should be from this._type

File: modules/sdk-coin-ton/src/lib/singleNominatorWithdrawBuilder.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import TonWeb from 'tonweb';
+import { Cell } from 'tonweb/dist/types/boc/cell';
+
+export class SingleNominatorWithdrawBuilder extends TransactionBuilder {
+  constructor(_coinConfig: Readonly<CoinConfig>) {
+    super(_coinConfig);
+  }
+
+  protected get transactionType(): TransactionType {
+    return TransactionType.StakingWithdraw;
```
Comment (ewangbitgo) [inline modules/sdk-coin-ton/src/lib/singleNominatorWithdrawBuilder.ts:13]: the type should be SingleNominatorWithdraw

File: modules/sdk-coin-ton/src/lib/singleNominatorWithdrawBuilder.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  }
+
+  protected get transactionType(): TransactionType {
+    return TransactionType.SingleNominatorWithdraw;
+  }
+
+  send(recipient: Recipient): SingleNominatorWithdrawBuilder {
+    this.transaction.recipient = recipient;
+    return this;
+  }
```
Comment (ewangbitgo) [inline modules/sdk-coin-ton/src/lib/singleNominatorWithdrawBuilder.ts:17]: you need add a method for withdrawAmount, and build message payload internally. And return error if user is trying to set a message.

File: modules/sdk-coin-ton/test/unit/singleNominatorWithdrawBuilder.ts [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+  const factory = new TransactionBuilderFactory(coins.get('tton'));
+  it('should build a unsigned withdraw tx', async function () {
+    const txId = 'wxaFRG6b7Fhw91Al3JY-zbvl9Mjr-P0NuLvoGYyS5FI='.replace(/\//g, '_').replace(/\+/g, '-');
+    const txBuilder = factory.getSingleNominatorWithdrawBuilder();
+    txBuilder.sender(testData.sender.address);
+    txBuilder.sequenceNumber(0);
+    txBuilder.publicKey(testData.sender.publicKey);
+    txBuilder.expireTime(1234567890);
+    txBuilder.send(testData.recipients[0]);
+    txBuilder.setWithdrawMessage('1');
```
Comment (ewangbitgo) [inline modules/sdk-coin-ton/test/unit/singleNominatorWithdrawBuilder.ts:20]: where to set withdraw amount

File: modules/sdk-coin-ton/test/unit/singleNominatorWithdrawBuilder.ts [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    txBuilder.bounceable(true);
+    const tx = await txBuilder.build();
+    should.equal(tx.type, TransactionType.SingleNominatorWithdraw);
+    should.equal(tx.toJson().bounceable, true);
```
Comment (ewangbitgo) [inline modules/sdk-coin-ton/test/unit/singleNominatorWithdrawBuilder.ts:56]: tx.toJson should have a withdraw amount decoded.

File: modules/sdk-coin-ton/test/unit/singleNominatorWithdrawBuilder.ts [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+    should.equal(builtTx2.type, TransactionType.SingleNominatorWithdraw);
+    should.equal(builtTx.toBroadcastFormat(), testData.signedTransferTransaction.tx);
+    builtTx.inputs.length.should.equal(1);
+    builtTx.outputs.length.should.equal(1);
+    jsonTx.sender.should.equal('EQAbJug-k-tufWMjEC1RKSM0iiJTDUcYkC7zWANHrkT55Fol');
+    jsonTx.destination.should.equal('EQA0i8-CdGnF_DhUHHf92R1ONH6sIA9vLZ_WLcCIhfBBXwtG');
+    jsonTx.amount.should.equal('10000000');
+    jsonTx.seqno.should.equal(6);
+    jsonTx.expirationTime.should.equal(1695997582);
+
```
Comment (ewangbitgo) [inline modules/sdk-coin-ton/test/unit/singleNominatorWithdrawBuilder.ts:95]: you should add a withdraw amount in toJson and verify in this test.

File: modules/sdk-coin-ton/test/unit/singleNominatorWithdrawBuilder.ts [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+    const txBuilderBounceable = factory.from(testData.signedTransferTransaction.txBounceable);
+
+    const tx = await txBuilder.build();
+    const txBounceable = await txBuilderBounceable.build();
+    tx.toJson().bounceable.should.equal(false);
+    txBounceable.toJson().bounceable.should.equal(true);
+  });
+
+  it('should build a signed withdraw tx and submit onchain', async function () {
+    const tonweb = new TonWeb(new TonWeb.HttpProvider('https://testnet.toncenter.com/api/v2/jsonRPC'));
```
Comment (ewangbitgo) [inline modules/sdk-coin-ton/test/unit/singleNominatorWithdrawBuilder.ts:122]: disable this test for ci

File: modules/sdk-coin-ton/test/unit/ton.ts [hunk 13]
```diff
@@ -175,11 +204,53 @@ describe('TON:', function () {
...
+        changeAmount: '0',
+        fee: { fee: 'UNKNOWN' },
+      });
+    });
+
+    it('should explain a single nominator withdraw transaction', async function () {
+      const explainedTransaction = (await basecoin.explainTransaction({
+        txHex: Buffer.from(testData.signedTransferTransaction.tx, 'base64').toString('hex'),
+      })) as TransactionExplanation;
+      explainedTransaction.should.deepEqual({
```
Comment (ewangbitgo) [inline modules/sdk-coin-ton/test/unit/ton.ts:222]: you need add withdraw amount to explainTransaction for single nominator transaction

File: modules/sdk-coin-ton/test/unit/singleNominatorWithdrawBuilder.ts (hunk 9 above)
Comment (quarterdill) [reply modules/sdk-coin-ton/test/unit/singleNominatorWithdrawBuilder.ts]: i split the message creation into `setWithdrawAmount` and `setWithdrawMessage` functions

File: modules/sdk-coin-ton/test/unit/singleNominatorWithdrawBuilder.ts (hunk 10 above)
Comment (quarterdill) [reply modules/sdk-coin-ton/test/unit/singleNominatorWithdrawBuilder.ts:59]: added the checks

**Review by ewangbitgo — APPROVED** [review]

**Pull request merged**